.github/workflows/publish.yml

```diff
@@ -8,18 +8,23 @@ on:
       - geo
 
 jobs:
+  e2e:
+    secrets: inherit
+    uses: ./.github/workflows/callable-e2e-tests.yml
+
   publish:
     name: Publish
     runs-on: ubuntu-latest
+    needs: [e2e]
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@24cb9080177205b6e8c946b17badbe402adc938f
 
       - name: Use Node.js 16 x64
-        uses: actions/setup-node@v2
+        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c
```

Comment:
Nit, but prefer if we add comment that indicates version number, so that it's easier to know when to upgrade.